File: android/src/main/kotlin/com/tradechart/plugin/engine/ChartEngine.kt
package com.tradechart.plugin.engine

import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.RectF
import android.view.Choreographer
import com.tradechart.plugin.bridge.ChartFlutterApiHolder
import com.tradechart.plugin.bridge.generated.CandleDataListMessage
import com.tradechart.plugin.bridge.generated.CandleDataMessage
import com.tradechart.plugin.bridge.generated.ChartInitParams
import com.tradechart.plugin.bridge.generated.ConfigMessage
import com.tradechart.plugin.bridge.generated.CrosshairDataMessage
import com.tradechart.plugin.bridge.generated.MarkerListMessage
import com.tradechart.plugin.bridge.generated.MarkerMessage
import com.tradechart.plugin.bridge.generated.ThemeMessage
import com.tradechart.plugin.bridge.generated.ViewportStateMessage
import com.tradechart.plugin.data.CandleStore
import com.tradechart.plugin.data.MarkerStore
import com.tradechart.plugin.renderer.AxisRenderer
import com.tradechart.plugin.renderer.BackgroundRenderer
import com.tradechart.plugin.renderer.CandleRenderer
import com.tradechart.plugin.renderer.ChartLayerRenderer
import com.tradechart.plugin.renderer.CrosshairRenderer
import com.tradechart.plugin.renderer.GridRenderer
import com.tradechart.plugin.renderer.LineRenderer
import com.tradechart.plugin.renderer.MarkerRenderer
import com.tradechart.plugin.renderer.VolumeRenderer
import com.tradechart.plugin.theme.NativeChartTheme
import com.tradechart.plugin.viewport.ViewportCalculator
import io.flutter.view.TextureRegistry

class ChartEngine(
    textureRegistry: TextureRegistry,
    private val flutterApiHolder: ChartFlutterApiHolder,
) {
    private val choreographer = Choreographer.getInstance()
    private val candleStore = CandleStore()
    private val markerStore = MarkerStore()
    private val viewport = ViewportCalculator()
    private val textureRenderer = TextureRenderer(textureRegistry)
    private val renderers: List<ChartLayerRenderer> = listOf(
        BackgroundRenderer(),
        GridRenderer(),
        VolumeRenderer(),
        CandleRenderer(),
        LineRenderer(),
        MarkerRenderer(),
        AxisRenderer(),
        CrosshairRenderer(),
    )

    private lateinit var theme: NativeChartTheme
    private lateinit var config: ConfigMessage
    private var chartType: String = "candle"
    private var width: Int = 1
    private var height: Int = 1
    private var scale: Float = 1f
    private var dirty: Boolean = false
    private var frameScheduled: Boolean = false
    private var lastFrameTimeNanos: Long = 0L
    private var viewportChangedPending: Boolean = false
    private var pendingCrosshairData: CrosshairDataMessage? = null
    private var crosshairState: CrosshairState? = null
    private var flingVelocityX: Double = 0.0
    private var flingActive: Boolean = false
    private var isDisposed: Boolean = false

    private var lastAnimFrameNanos: Long = 0L

    private val frameCallback = Choreographer.FrameCallback { frameTimeNanos ->
        onFrame(frameTimeNanos)
    }

    val textureId: Long
        get() = textureRenderer.textureId

    fun initialize(params: ChartInitParams) {
        width = params.width.toInt().coerceAtLeast(1)
        height = params.height.toInt().coerceAtLeast(1)
        scale = params.devicePixelRatio.toFloat().coerceAtLeast(1f)
        theme = NativeChartTheme(params.theme)
        config = params.config
        chartType = params.config.initialChartType
        textureRenderer.resize(width, height, scale)
        dirty = true
        renderCurrentFrame()
        flutterApiHolder.onChartReady()
        flutterApiHolder.onViewportChanged(viewportState())
    }

    fun dispose() {
        isDisposed = true
        if (frameScheduled) {
            choreographer.removeFrameCallback(frameCallback)
            frameScheduled = false
        }
        textureRenderer.release()
    }

    fun onSizeChanged(width: Double, height: Double) {
        this.width = width.toInt().coerceAtLeast(1)
        this.height = height.toInt().coerceAtLeast(1)
        textureRenderer.resize(this.width, this.height, scale)
        recomputeViewport()
        markDirty(viewportChanged = candleStore.count > 0)
    }

    fun loadCandles(data: CandleDataListMessage) {
        candleStore.load(data)
        recomputeViewport()
        markDirty(viewportChanged = true)
    }

    fun appendCandle(candle: CandleDataMessage) {
        val previousCount = candleStore.count
        val lastTimestamp = candleStore.lastTimestampOrNull()
        if (lastTimestamp != null && candle.timestamp <= lastTimestamp) {
            return
        }
        val wasAtLatest = viewport.isAtLatest
        candleStore.append(candle)
        recomputeViewportForAppend(wasAtLatest, previousCount == 0)
        val viewportChanged = previousCount == 0 || (wasAtLatest && config.autoScrollOnAppend)
        markDirty(viewportChanged = viewportChanged)
    }

    fun updateLastCandle(candle: CandleDataMessage) {
        val lastTimestamp = candleStore.lastTimestampOrNull()
        if (lastTimestamp != null && candle.timestamp != lastTimestamp) {
            return
        }
        val lastWasVisible = candleStore.count == 0 ||
            (viewport.hasVisibleRange() && viewport.visibleEndIndex == candleStore.count - 1)
        candleStore.updateLast(candle)
        if (lastWasVisible) {
            recomputeViewport()
        }
        markDirty(viewportChanged = lastWasVisible)
    }

    fun setMarkers(markers: MarkerListMessage) {
        if (markers.markers.isEmpty() && candleStore.count == 0) {
            markerStore.set(emptyList())
            return
        }
        markerStore.set(markers.markers)
        markDirty()
    }

    fun addMarker(marker: MarkerMessage) {
        markerStore.add(marker)
        markDirty()
    }

    fun clearMarkers() {
        if (candleStore.count == 0) {
            markerStore.clear()
            return
        }
        markerStore.clear()
        markDirty()
    }

    fun setChartType(chartType: String) {
        this.chartType = chartType
        recomputeViewport()
        markDirty(viewportChanged = candleStore.count > 0)
    }

    fun setTimeframe(timeframe: String) {
        candleStore.clear()
        crosshairState = null
        pendingCrosshairData = null
        endFling()
        recomputeViewport()
        dirty = true
        renderCurrentFrame()
    }

    fun setTheme(theme: ThemeMessage) {
        this.theme = NativeChartTheme(theme)
        markDirty()
    }

    fun setConfig(config: ConfigMessage) {
        this.config = config
        chartType = chartType.ifBlank { config.initialChartType }
        recomputeViewport()
        markDirty(viewportChanged = candleStore.count > 0)
    }

    fun scrollToEnd() {
        endFling()
        if (viewport.scrollToEnd(candleStore, plotWidth(), config, chartType)) {
            markDirty(viewportChanged = true)
        }
    }

    fun onPanUpdate(deltaX: Double) {
        endFling()
        if (viewport.applyPanDelta(deltaX, candleStore, plotWidth(), config, chartType)) {
            markDirty(viewportChanged = true)
        }
    }

    fun onPanEnd(velocityX: Double) {
        if (kotlin.math.abs(velocityX) < 50.0 || candleStore.count == 0) {
            endFling()
            return
        }
        flingVelocityX = velocityX
        flingActive = true
        lastFrameTimeNanos = 0L
        scheduleFrame()
    }

    fun onScaleUpdate(scaleFactor: Double, focalPointX: Double) {
        endFling()
        if (viewport.applyScale(scaleFactor, focalPointX, candleStore, plotWidth(), config, chartType)) {
            markDirty(viewportChanged = true)
        }
    }

    fun onScaleEnd() {
        viewport.endScaleGesture()
    }

    fun onCrosshairStart(x: Double, y: Double) {
        if (!config.enableCrosshair) {
            return
        }
        val layout = createLayout()
        val resolved = viewport.resolveCrosshair(
            x = x,
            y = y,
            contentRect = layout.contentRect,
            mainChartRect = layout.mainChartRect,
            candleStore = candleStore,
        ) ?: return
        crosshairState = CrosshairState(
            candleIndex = resolved.candleIndex,
            snappedX = resolved.snappedX,
            y = resolved.y,
        )
        pendingCrosshairData = crosshairDataMessage(crosshairState!!)
        markDirty()
    }

    fun onCrosshairMove(x: Double, y: Double) {
        if (!config.enableCrosshair) {
            return
        }
        val layout = createLayout()
        val resolved = viewport.resolveCrosshair(
            x = x,
            y = y,
            contentRect = layout.contentRect,
            mainChartRect = layout.mainChartRect,
            candleStore = candleStore,
        ) ?: return
        crosshairState = CrosshairState(
            candleIndex = resolved.candleIndex,
            snappedX = resolved.snappedX,
            y = resolved.y,
        )
        pendingCrosshairData = crosshairDataMessage(crosshairState!!)
        markDirty()
    }

    fun onCrosshairEnd() {
        if (crosshairState == null && pendingCrosshairData == null) {
            return
        }
        crosshairState = null
        pendingCrosshairData = null
        markDirty()
    }

    private fun recomputeViewport() {
        val layout = createLayout()
        viewport.refresh(
            candleStore = candleStore,
            plotWidth = layout.mainChartRect.width().toDouble(),
            config = config,
            chartType = chartType,
        )
    }

    private fun recomputeViewportForAppend(
        wasAtLatest: Boolean,
        isFirstCandle: Boolean,
    ) {
        if (isFirstCandle) {
            recomputeViewport()
            return
        }
        if (wasAtLatest && config.autoScrollOnAppend) {
            viewport.scrollToEnd(candleStore, plotWidth(), config, chartType)
            return
        }
        recomputeViewport()
    }

    private fun renderCurrentFrame() {
        if (isDisposed) {
            return
        }
        val layout = createLayout()
        if (candleStore.count > 0) {
            viewport.refresh(
                candleStore = candleStore,
                plotWidth = layout.mainChartRect.width().toDouble(),
                config = config,
                chartType = chartType,
            )
        }

        val renderChartType = viewport.effectiveRenderChartType(chartType)
        textureRenderer.render { canvas ->
            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
            val frame = ChartFrame(
                canvas = canvas,
                candleStore = candleStore,
                markerStore = markerStore,
                viewport = viewport,
                theme = theme,
                config = config,
                chartType = renderChartType,
                width = width,
                height = height,
                contentRect = layout.contentRect,
                mainChartRect = layout.mainChartRect,
                volumeRect = layout.volumeRect,
                priceAxisRect = layout.priceAxisRect,
                timeAxisRect = layout.timeAxisRect,
                crosshair = crosshairState,
            )
            renderers.forEach { renderer ->
                renderer.render(frame)
            }
        }
    }

    private fun viewportState(): ViewportStateMessage {
        if (candleStore.count == 0 || !viewport.hasVisibleRange()) {
            return ViewportStateMessage(
                startTimestamp = 0,
                endTimestamp = 0,
                priceHigh = 0.0,
                priceLow = 0.0,
                visibleCandleCount = 0,
                candleWidth = 0.0,
                isAtLatest = true,
            )
        }
        return ViewportStateMessage(
            startTimestamp = candleStore.timestampAt(viewport.visibleStartIndex),
            endTimestamp = candleStore.timestampAt(viewport.visibleEndIndex),
            priceHigh = viewport.priceHigh,
            priceLow = viewport.priceLow,
            visibleCandleCount = viewport.visibleCandleCount.toLong(),
            candleWidth = viewport.candleWidth,
            isAtLatest = viewport.isAtLatest,
        )
    }

    private fun markDirty(
        viewportChanged: Boolean = false,
    ) {
        dirty = true
        if (viewportChanged) {
            viewportChangedPending = true
        }
        scheduleFrame()
    }

    private fun scheduleFrame() {
        if (frameScheduled || isDisposed) {
            return
        }
        frameScheduled = true
        choreographer.postFrameCallback(frameCallback)
    }

    private fun onFrame(frameTimeNanos: Long) {
        frameScheduled = false
        if (isDisposed) {
            return
        }

        if (flingActive) {
            updateFling(frameTimeNanos)
        } else {
            lastFrameTimeNanos = 0L
        }

        if (candleStore.isAnimating) {
            val dt = if (lastAnimFrameNanos == 0L) {
                1.0 / 60.0
            } else {
                (frameTimeNanos - lastAnimFrameNanos) / 1_000_000_000.0
            }
            lastAnimFrameNanos = frameTimeNanos
            candleStore.advanceAnimation(dt)
            dirty = true
        } else {
            lastAnimFrameNanos = 0L
        }

        if (dirty) {
            renderCurrentFrame()
            dirty = false
            if (viewportChangedPending) {
                flutterApiHolder.onViewportChanged(viewportState())
                viewportChangedPending = false
            }
            pendingCrosshairData?.let {
                flutterApiHolder.onCrosshairData(it)
                pendingCrosshairData = null
            }
        }

        if (flingActive || candleStore.isAnimating || dirty) {
            scheduleFrame()
        }
    }

    private fun updateFling(frameTimeNanos: Long) {
        if (lastFrameTimeNanos == 0L) {
            lastFrameTimeNanos = frameTimeNanos
            return
        }
        val dtSeconds = (frameTimeNanos - lastFrameTimeNanos) / 1_000_000_000.0
        lastFrameTimeNanos = frameTimeNanos
        if (dtSeconds <= 0.0) {
            return
        }
        val deltaX = flingVelocityX * dtSeconds
        val moved = viewport.applyPanDelta(deltaX, candleStore, plotWidth(), config, chartType)
        if (moved) {
            dirty = true
            viewportChangedPending = true
        }
        flingVelocityX *= kotlin.math.exp(-6.0 * dtSeconds)
        if (kotlin.math.abs(flingVelocityX) < 10.0 || !moved) {
            endFling()
        }
    }

    private fun endFling() {
        flingActive = false
        flingVelocityX = 0.0
        lastFrameTimeNanos = 0L
    }

    private fun plotWidth(): Double = createLayout().mainChartRect.width().toDouble()

    private fun crosshairDataMessage(crosshair: CrosshairState): CrosshairDataMessage {
        return CrosshairDataMessage(
            timestamp = candleStore.timestampAt(crosshair.candleIndex),
            open = candleStore.openAt(crosshair.candleIndex),
            high = candleStore.highAt(crosshair.candleIndex),
            low = candleStore.lowAt(crosshair.candleIndex),
            close = candleStore.closeAt(crosshair.candleIndex),
            volume = candleStore.volumeAt(crosshair.candleIndex),
            x = crosshair.snappedX.toDouble(),
            y = crosshair.y.toDouble(),
        )
    }

    private fun createLayout(): Layout {
        val axisWidth = if (config.showAxis) 56f else 0f
        val timeAxisHeight = if (config.showAxis) 24f else 0f
        val contentRight = width.toFloat() - axisWidth
        val contentBottom = height.toFloat() - timeAxisHeight
        val contentRect = RectF(0f, 0f, contentRight, contentBottom)

        val volumeGap = if (config.showVolume) 8f else 0f
        val volumeHeight = if (config.showVolume) {
            (contentRect.height() * config.volumeHeightRatio.toFloat()).coerceAtLeast(48f)
        } else {
            0f
        }
        val mainBottom = contentRect.bottom - volumeHeight - volumeGap
        val mainChartRect = RectF(
            contentRect.left,
            contentRect.top,
            contentRect.right,
            if (config.showVolume) mainBottom else contentRect.bottom,
        )
        val volumeRect = if (config.showVolume) {
            RectF(
                contentRect.left,
                mainChartRect.bottom + volumeGap,
                contentRect.right,
                contentRect.bottom,
            )
        } else {
            null
        }
        val priceAxisRect = if (config.showAxis) {
            RectF(contentRect.right, 0f, width.toFloat(), mainChartRect.bottom)
        } else {
            null
        }
        val timeAxisRect = if (config.showAxis) {
            RectF(0f, contentBottom, width.toFloat(), height.toFloat())
        } else {
            null
        }

        return Layout(
            contentRect = contentRect,
            mainChartRect = mainChartRect,
            volumeRect = volumeRect,
            priceAxisRect = priceAxisRect,
            timeAxisRect = timeAxisRect,
        )
    }

    private data class Layout(
        val contentRect: RectF,
        val mainChartRect: RectF,
        val volumeRect: RectF?,
        val priceAxisRect: RectF?,
        val timeAxisRect: RectF?,
    )
}

File: ios/Classes/Engine/ChartEngine.swift
import CoreGraphics
import Flutter
import Foundation
import QuartzCore

final class ChartEngine: NSObject {
  init(chartId: Int64, textureRegistry: FlutterTextureRegistry, flutterApiHolder: ChartFlutterApiHolder) {
    self.chartId = chartId
    self.flutterApiHolder = flutterApiHolder
    textureRenderer = TextureRenderer(textureRegistry: textureRegistry)
    candleStore = CandleStore()
    markerStore = MarkerStore()
    viewport = ViewportCalculator()
    renderers = [
      BackgroundRenderer(),
      GridRenderer(),
      VolumeRenderer(),
      CandleRenderer(),
      LineRenderer(),
      MarkerRenderer(),
      AxisRenderer(),
      CrosshairRenderer(),
    ]
    super.init()
  }

  private let chartId: Int64
  private let flutterApiHolder: ChartFlutterApiHolder
  private let textureRenderer: TextureRenderer
  private let candleStore: CandleStore
  private let markerStore: MarkerStore
  private let viewport: ViewportCalculator
  private let renderers: [ChartLayerRenderer]

  private var theme: NativeChartTheme?
  private var config: ConfigMessage?
  private var chartType: String = "candle"
  private var width: Int = 1
  private var height: Int = 1
  private var scale: CGFloat = 1.0
  private var dirty: Bool = false
  private var frameScheduled: Bool = false
  private var lastFrameTime: CFTimeInterval = 0
  private var viewportChangedPending: Bool = false
  private var pendingCrosshairData: CrosshairDataMessage?
  private var crosshairState: CrosshairState?
  private var flingVelocityX: Double = 0
  private var flingActive: Bool = false
  private var isDisposed: Bool = false
  private var displayLink: CADisplayLink?

  var textureId: Int64 {
    textureRenderer.textureId
  }

  func initialize(params: ChartInitParams) {
    width = max(Int(params.width), 1)
    height = max(Int(params.height), 1)
    scale = max(CGFloat(params.devicePixelRatio), 1.0)
    theme = NativeChartTheme(message: params.theme)
    config = params.config
    chartType = params.config.initialChartType
    textureRenderer.resize(width: width, height: height, scale: scale)
    dirty = true
    renderCurrentFrame()
    flutterApiHolder.onChartReady(chartId: chartId)
    flutterApiHolder.onViewportChanged(chartId: chartId, viewport: viewportState())
  }

  func dispose() {
    isDisposed = true
    invalidateDisplayLink()
    textureRenderer.releaseTexture()
  }

  func onSizeChanged(width: Double, height: Double) {
    self.width = max(Int(width), 1)
    self.height = max(Int(height), 1)
    textureRenderer.resize(width: self.width, height: self.height, scale: scale)
    recomputeViewport()
    markDirty(viewportChanged: candleStore.count > 0)
  }

  func loadCandles(data: CandleDataListMessage) {
    candleStore.load(data: data)
    recomputeViewport()
    markDirty(viewportChanged: true)
  }

  func appendCandle(candle: CandleDataMessage) {
    let previousCount = candleStore.count
    if let lastTimestamp = candleStore.lastTimestampOrNil(), candle.timestamp <= lastTimestamp {
      return
    }
    let wasAtLatest = viewport.isAtLatest
    candleStore.append(candle: candle)
    recomputeViewportForAppend(wasAtLatest: wasAtLatest, isFirstCandle: previousCount == 0)
    let viewportChanged = previousCount == 0 || (wasAtLatest && (config?.autoScrollOnAppend ?? false))
    markDirty(viewportChanged: viewportChanged)
  }

  func updateLastCandle(candle: CandleDataMessage) {
    if let lastTimestamp = candleStore.lastTimestampOrNil(), candle.timestamp != lastTimestamp {
      return
    }
    let lastWasVisible = candleStore.count == 0 ||
      (viewport.hasVisibleRange() && viewport.visibleEndIndex == candleStore.count - 1)
    candleStore.updateLast(candle: candle)
    if lastWasVisible {
      recomputeViewport()
    }
    markDirty(viewportChanged: lastWasVisible)
  }

  func setMarkers(markers: MarkerListMessage) {
    if markers.markers.isEmpty && candleStore.count == 0 {
      markerStore.set(markers: [])
      return
    }
    markerStore.set(markers: markers.markers)
    markDirty()
  }

  func addMarker(marker: MarkerMessage) {
    markerStore.add(marker: marker)
    markDirty()
  }

  func clearMarkers() {
    if candleStore.count == 0 {
      markerStore.clear()
      return
    }
    markerStore.clear()
    markDirty()
  }

  func setChartType(chartType: String) {
    self.chartType = chartType
    recomputeViewport()
    markDirty(viewportChanged: candleStore.count > 0)
  }

  func setTimeframe(timeframe: String) {
    candleStore.clear()
    crosshairState = nil
    pendingCrosshairData = nil
    endFling()
    recomputeViewport()
    dirty = true
    renderCurrentFrame()
    dirty = false
    flutterApiHolder.onViewportChanged(chartId: chartId, viewport: viewportState())
  }

  func setTheme(theme: ThemeMessage) {
    self.theme = NativeChartTheme(message: theme)
    markDirty()
  }

  func setConfig(config: ConfigMessage) {
    self.config = config
    chartType = chartType.isEmpty ? config.initialChartType : chartType
    recomputeViewport()
    markDirty(viewportChanged: candleStore.count > 0)
  }

  func scrollToEnd() {
    guard let config else { return }
    endFling()
    if viewport.scrollToEnd(
      candleStore: candleStore,
      plotWidth: plotWidth(),
      config: config,
      chartType: chartType
    ) {
      markDirty(viewportChanged: true)
    }
  }

  func onPanUpdate(deltaX: Double) {
    guard let config else { return }
    endFling()
    if viewport.applyPanDelta(
      deltaX: deltaX,
      candleStore: candleStore,
      plotWidth: plotWidth(),
      config: config,
      chartType: chartType
    ) {
      markDirty(viewportChanged: true)
    }
  }

  func onPanEnd(velocityX: Double) {
    guard abs(velocityX) >= 50, candleStore.count > 0 else {
      endFling()
      return
    }
    flingVelocityX = velocityX
    flingActive = true
    lastFrameTime = 0
    scheduleFrame()
  }

  func onScaleUpdate(scaleFactor: Double, focalPointX: Double) {
    guard let config else { return }
    endFling()
    if viewport.applyScale(
      scaleFactor: scaleFactor,
      focalPointX: focalPointX,
      candleStore: candleStore,
      plotWidth: plotWidth(),
      config: config,
      chartType: chartType
    ) {
      markDirty(viewportChanged: true)
    }
  }

  func onScaleEnd() {
    viewport.endScaleGesture()
  }

  func onCrosshairStart(x: Double, y: Double) {
    guard config?.enableCrosshair != false else { return }
    let layout = createLayout()
    guard let resolved = viewport.resolveCrosshair(
      x: x,
      y: y,
      contentRect: layout.contentRect,
      mainChartRect: layout.mainChartRect,
      candleStore: candleStore
    ) else {
      return
    }
    crosshairState = CrosshairState(
      candleIndex: resolved.candleIndex,
      snappedX: resolved.snappedX,
      y: resolved.y
    )
    pendingCrosshairData = crosshairDataMessage(crosshairState!)
    markDirty()
  }

  func onCrosshairMove(x: Double, y: Double) {
    guard config?.enableCrosshair != false else { return }
    let layout = createLayout()
    guard let resolved = viewport.resolveCrosshair(
      x: x,
      y: y,
      contentRect: layout.contentRect,
      mainChartRect: layout.mainChartRect,
      candleStore: candleStore
    ) else {
      return
    }
    crosshairState = CrosshairState(
      candleIndex: resolved.candleIndex,
      snappedX: resolved.snappedX,
      y: resolved.y
    )
    pendingCrosshairData = crosshairDataMessage(crosshairState!)
    markDirty()
  }

  func onCrosshairEnd() {
    guard crosshairState != nil || pendingCrosshairData != nil else { return }
    crosshairState = nil
    pendingCrosshairData = nil
    markDirty()
  }

  private func recomputeViewport() {
    guard let config else { return }
    let layout = createLayout()
    viewport.refresh(
      candleStore: candleStore,
      plotWidth: layout.mainChartRect.width,
      config: config,
      chartType: chartType
    )
  }

  private func recomputeViewportForAppend(wasAtLatest: Bool, isFirstCandle: Bool) {
    guard let config else { return }
    if isFirstCandle {
      recomputeViewport()
      return
    }
    if wasAtLatest && config.autoScrollOnAppend {
      _ = viewport.scrollToEnd(
        candleStore: candleStore,
        plotWidth: plotWidth(),
        config: config,
        chartType: chartType
      )
      return
    }
    recomputeViewport()
  }

  private func renderCurrentFrame() {
    guard let theme, let config, !isDisposed else { return }
    let layout = createLayout()
    if candleStore.count > 0 {
      viewport.refresh(
        candleStore: candleStore,
        plotWidth: layout.mainChartRect.width,
        config: config,
        chartType: chartType
      )
    }

    let renderChartType = viewport.effectiveRenderChartType(userChartType: chartType)
    textureRenderer.render { context in
      let frame = ChartFrame(
        context: context,
        candleStore: candleStore,
        markerStore: markerStore,
        viewport: viewport,
        theme: theme,
        config: config,
        chartType: renderChartType,
        width: width,
        height: height,
        contentRect: layout.contentRect,
        mainChartRect: layout.mainChartRect,
        volumeRect: layout.volumeRect,
        priceAxisRect: layout.priceAxisRect,
        timeAxisRect: layout.timeAxisRect,
        crosshair: crosshairState
      )
      renderers.forEach { $0.render(frame: frame) }
    }
  }

  private func viewportState() -> ViewportStateMessage {
    guard candleStore.count > 0, viewport.hasVisibleRange() else {
      return ViewportStateMessage(
        startTimestamp: 0,
        endTimestamp: 0,
        priceHigh: 0,
        priceLow: 0,
        visibleCandleCount: 0,
        candleWidth: 0,
        isAtLatest: true
      )
    }
    return ViewportStateMessage(
      startTimestamp: candleStore.timestamp(at: viewport.visibleStartIndex),
      endTimestamp: candleStore.timestamp(at: viewport.visibleEndIndex),
      priceHigh: viewport.priceHigh,
      priceLow: viewport.priceLow,
      visibleCandleCount: Int64(viewport.visibleCandleCount),
      candleWidth: viewport.candleWidth,
      isAtLatest: viewport.isAtLatest
    )
  }

  private func markDirty(viewportChanged: Bool = false) {
    dirty = true
    if viewportChanged {
      viewportChangedPending = true
    }
    scheduleFrame()
  }

  private func scheduleFrame() {
    guard !frameScheduled, !isDisposed else { return }
    if displayLink == nil {
      let link = CADisplayLink(target: self, selector: #selector(handleDisplayLink(_:)))
      if #available(iOS 15.0, *) {
        link.preferredFrameRateRange = CAFrameRateRange(minimum: 30, maximum: 60, preferred: 60)
      }
      link.add(to: .main, forMode: .common)
      displayLink = link
    }
    frameScheduled = true
    displayLink?.isPaused = false
  }

  @objc
  private func handleDisplayLink(_ link: CADisplayLink) {
    frameScheduled = false
    if isDisposed {
      invalidateDisplayLink()
      return
    }

    if flingActive {
      updateFling(link.targetTimestamp)
    } else {
      lastFrameTime = 0
    }

    if candleStore.isAnimating {
      let dt = link.duration > 0 ? link.duration : 1.0 / 60.0
      candleStore.advanceAnimation(dt: dt)
      dirty = true
    }

    if dirty {
      renderCurrentFrame()
      dirty = false
      if viewportChangedPending {
        flutterApiHolder.onViewportChanged(chartId: chartId, viewport: viewportState())
        viewportChangedPending = false
      }
      if let pendingCrosshairData {
        flutterApiHolder.onCrosshairData(chartId: chartId, data: pendingCrosshairData)
        self.pendingCrosshairData = nil
      }
    }

    if flingActive || candleStore.isAnimating || dirty {
      frameScheduled = true
      return
    }

    displayLink?.isPaused = true
  }

  private func updateFling(_ frameTime: CFTimeInterval) {
    guard let config else { return }
    if lastFrameTime == 0 {
      lastFrameTime = frameTime
      return
    }
    let dt = frameTime - lastFrameTime
    lastFrameTime = frameTime
    guard dt > 0 else { return }

    let deltaX = flingVelocityX * dt
    let moved = viewport.applyPanDelta(
      deltaX: deltaX,
      candleStore: candleStore,
      plotWidth: plotWidth(),
      config: config,
      chartType: chartType
    )
    if moved {
      dirty = true
      viewportChangedPending = true
    }
    flingVelocityX *= Foundation.exp(-6 * dt)
    if abs(flingVelocityX) < 10 || !moved {
      endFling()
    }
  }

  private func endFling() {
    flingActive = false
    flingVelocityX = 0
    lastFrameTime = 0
  }

  private func plotWidth() -> Double {
    createLayout().mainChartRect.width
  }

  private func crosshairDataMessage(_ crosshair: CrosshairState) -> CrosshairDataMessage {
    CrosshairDataMessage(
      timestamp: candleStore.timestamp(at: crosshair.candleIndex),
      open: candleStore.open(at: crosshair.candleIndex),
      high: candleStore.high(at: crosshair.candleIndex),
      low: candleStore.low(at: crosshair.candleIndex),
      close: candleStore.close(at: crosshair.candleIndex),
      volume: candleStore.volume(at: crosshair.candleIndex),
      x: Double(crosshair.snappedX),
      y: Double(crosshair.y)
    )
  }

  private func invalidateDisplayLink() {
    displayLink?.invalidate()
    displayLink = nil
    frameScheduled = false
  }

  private func createLayout() -> Layout {
    guard let config else {
      let rect = CGRect(x: 0, y: 0, width: width, height: height)
      return Layout(
        contentRect: rect,
        mainChartRect: rect,
        volumeRect: nil,
        priceAxisRect: nil,
        timeAxisRect: nil
      )
    }

    let axisWidth: CGFloat = config.showAxis ? 56 : 0
    let timeAxisHeight: CGFloat = config.showAxis ? 24 : 0
    let contentRight = CGFloat(width) - axisWidth
    let contentBottom = CGFloat(height) - timeAxisHeight
    let contentRect = CGRect(x: 0, y: 0, width: contentRight, height: contentBottom)

    let volumeGap: CGFloat = config.showVolume ? 8 : 0
    let volumeHeight: CGFloat = config.showVolume
      ? max(48, contentRect.height * config.volumeHeightRatio)
      : 0
    let mainBottom = contentRect.maxY - volumeHeight - volumeGap
    let mainChartRect = CGRect(
      x: contentRect.minX,
      y: contentRect.minY,
      width: contentRect.width,
      height: config.showVolume ? max(32, mainBottom - contentRect.minY) : contentRect.height
    )
    let volumeRect = config.showVolume
      ? CGRect(
        x: contentRect.minX,
        y: mainChartRect.maxY + volumeGap,
        width: contentRect.width,
        height: max(0, contentRect.maxY - (mainChartRect.maxY + volumeGap))
      )
      : nil
    let priceAxisRect = config.showAxis
      ? CGRect(x: contentRect.maxX, y: 0, width: CGFloat(width) - contentRect.maxX, height: mainChartRect.maxY)
      : nil
    let timeAxisRect = config.showAxis
      ? CGRect(x: 0, y: contentBottom, width: CGFloat(width), height: CGFloat(height) - contentBottom)
      : nil

    return Layout(
      contentRect: contentRect,
      mainChartRect: mainChartRect,
      volumeRect: volumeRect,
      priceAxisRect: priceAxisRect,
      timeAxisRect: timeAxisRect
    )
  }

  private struct Layout {
    let contentRect: CGRect
    let mainChartRect: CGRect
    let volumeRect: CGRect?
    let priceAxisRect: CGRect?
    let timeAxisRect: CGRect?
  }
}
